File: Objective-C/EventCapture/EventCapture/RunloopEventCapture.m
//
//  MyRunLoop.m
//  EventCapture
//
//  Created by zidon on 16/8/17.
//  Copyright © 2016年 zidon. All rights reserved.
//

#import "RunloopEventCapture.h"
#import <dlfcn.h>
#import "fishhook.h"

@interface RunloopEventCapture ()

@property (nonatomic, strong) NSMutableArray *tasks;

@property (nonatomic, strong) NSMutableArray *tasksKeys;

@property (nonatomic, strong) NSTimer *timer;

@end

@implementation RunloopEventCapture
/**
 *  添加单个事件对象监听
 *
 *  @param unit 回调block
 *  @param key  唯一key值
 */
- (void)addTask:(MyRunLoopWorkDistributionUnit)unit withKey:(id)key{
    [self.tasks addObject:unit];
    [self.tasksKeys addObject:key];
    if (self.tasks.count > self.maxiTasks) {
        [self.tasks removeObjectAtIndex:0];
        [self.tasksKeys removeObjectAtIndex:0];
    }
}
/**
 *  同时添加多个事件监听对象
 *
 *  @param keys key值的数组
 */
- (void)addTaskWithKeys:(NSArray *)keys {
    [self.tasksKeys addObjectsFromArray:keys];
    if (self.tasks.count > self.maxiTasks) {
        [self.tasks removeObjectAtIndex:0];
        [self.tasksKeys removeObjectAtIndex:0];
    }
}
/**
 *  同时添加多个事件监听对象
 *
 *  @param items 模型的数组
 */
- (void)addTaskWithItems:(NSArray *)items {
    
    [self.tasksKeys addObjectsFromArray:items];
    if (self.tasks.count > self.maxiTasks) {
        [self.tasks removeObjectAtIndex:0];
        [self.tasksKeys removeObjectAtIndex:0];
    }
}

- (instancetype)init {
    
    if ((self = [super init])) {
        _maxiTasks = 30;
        _tasks = [NSMutableArray array];
        _tasksKeys = [NSMutableArray array];
        //_timer = [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(_timerFiredMethod:) userInfo:nil repeats:YES];
    }
    return self;
}

+ (instancetype)sharedRunLoopWorkDistribution {
    static RunloopEventCapture *singleton;
    static dispatch_once_t once;
    dispatch_once(&once, ^{
        singleton = [[RunloopEventCapture alloc] init];
        [self _registerRunLoopWorkDistributionAsMainRunloopObserver:singleton];
    });
    return singleton;
}

+ (void)_registerRunLoopWorkDistributionAsMainRunloopObserver:(RunloopEventCapture *)runLoopWorkDistribution {
    static CFRunLoopObserverRef defaultModeObserver;
    //_defaultModeRunLoopWorkDistributionCallback 方法地址
    _registerObserver(kCFRunLoopBeforeWaiting, defaultModeObserver, NSIntegerMax - 999, kCFRunLoopDefaultMode, (__bridge void *)runLoopWorkDistribution, &_defaultModeRunLoopWorkDistributionCallback);
}

//监听里存了方法的地址_defaultModeRunLoopWorkDistributionCallback
static void _defaultModeRunLoopWorkDistributionCallback(CFRunLoopObserverRef observer, CFRunLoopActivity activity, void *info) {
    _runLoopWorkDistributionCallback(observer, activity, info);
}

static void _runLoopWorkDistributionCallback(CFRunLoopObserverRef observer, CFRunLoopActivity activity, void *info) {
    
    RunloopEventCapture *runLoopWorkDistribution = (__bridge RunloopEventCapture *)info;
    
    NSLog(@":::%@",printActivity(activity));
    
    if (runLoopWorkDistribution.tasks.count == 0) {
        return;
    }
    BOOL result = NO;
    while (result == NO && runLoopWorkDistribution.tasks.count) {
        MyRunLoopWorkDistributionUnit unit  = runLoopWorkDistribution.tasks.firstObject;
        if ([runLoopWorkDistribution.tasksKeys[0] isEqualToString:@"lala"]) {
            
            if (runLoopWorkDistribution.delegate&&[runLoopWorkDistribution.delegate respondsToSelector:@selector(testUploadCaptureUserData:)]) {
                /**
                 *  需要埋点的记录，异步上传
                 */
                dispatch_async(dispatch_get_global_queue(0, 0), ^{
                    [runLoopWorkDistribution.delegate testUploadCaptureUserData:@"对头，要的就是你"];
                });
            }
        }
        result = unit();
        [runLoopWorkDistribution.tasks removeObjectAtIndex:0];
        [runLoopWorkDistribution.tasksKeys removeObjectAtIndex:0];
    }
}

static inline NSString *printActivity(CFRunLoopActivity activity) {
    
    NSString *activityDescription;
    switch (activity) {
        case kCFRunLoopEntry:
            activityDescription = @"kCFRunLoopEntry";
            break;
        case kCFRunLoopBeforeTimers:
            activityDescription = @"kCFRunLoopBeforeTimers";
            break;
        case kCFRunLoopBeforeSources:
            activityDescription = @"kCFRunLoopBeforeSources";
            break;
        case kCFRunLoopBeforeWaiting:
            activityDescription = @"kCFRunLoopBeforeWaiting";
            break;
        case kCFRunLoopAfterWaiting:
            activityDescription = @"kCFRunLoopAfterWaiting";
            break;
        case kCFRunLoopExit:
            activityDescription = @"kCFRunLoopExit";
            break;
        default:
            break;
            
    }
    return activityDescription;
}

static void _registerObserver(CFOptionFlags activities, CFRunLoopObserverRef observer, CFIndex order, CFStringRef mode, void *info, CFRunLoopObserverCallBack callback) {
    
    CFRunLoopRef runLoop = CFRunLoopGetCurrent();
    CFRunLoopObserverContext context = {
        0,
        info,
        &CFRetain,
        &CFRelease,
        NULL
    };
    observer = CFRunLoopObserverCreate(NULL,
                                       activities,
                                       YES,
                                       order,
                                       callback,
                                       &context);
    
    CFRunLoopAddObserver(runLoop, observer, mode);
    CFRelease(observer);
}

static void (*orig_CFRunLoopAddObserver)(CFRunLoopRef, CFRunLoopObserverRef, CFRunLoopMode);
void my_CFRunLoopAddObserver(CFRunLoopRef rl, CFRunLoopObserverRef observer, CFRunLoopMode mode) {
    NSLog(@"\n---------:Calling CFRunLoopAddObserver(runloop(%p), %@, %@)\n\n",
          rl,
          observer,
          mode);
    orig_CFRunLoopAddObserver(rl, observer, mode);
}

static struct rebinding rebindings[] = {
    { "CFRunLoopAddObserver",
        my_CFRunLoopAddObserver,
        (void *)&orig_CFRunLoopAddObserver },
};

void inception_runloop() {
    rebind_symbols(rebindings, sizeof(rebindings)/sizeof(struct rebinding));
}

@end

File: Swift/SwiftEventCapture/SwiftEventCapture/RunloopEventCapture.swift
//
//  RunloopEventCapture.swift
//  SwiftEventCapture
//
//  Created by zidon on 16/8/17.
//  Copyright © 2016年 zidon. All rights reserved.
//

import Foundation

@objc protocol RunloopEventCaptureDelegate {
    //返回值代表继续执行事件（已经添加的事件）
    @objc optional func asyncUploadCaptureUserData(data:AnyObject) -> Bool
}

typealias MyRunLoopWorkDistributionUnit = (Bool) -> Void

class RunloopEventCapture: NSObject {
    
    var captureObjct:AnyObject?
    
    override init() {
        super.init()
        resgistRunloopWork(obj: self)
    }
    
    static let sharedInstace=RunloopEventCapture()
    
    /**
     添加单个事件对象监听,在事件的触发处调用这个方法
     
     - parameter key:  回调的block
     - parameter unit: 唯一的key值
     */
    func addTask(key:AnyObject) {
        
        captureObjct=key
        
        //可以设置最大的任务数量
        if self.taskContent.count > self.maxTasks {
            self.taskContent.removeAll()
        }
    }
    
    /**
     添加多个事件对象的监听,在事件的触发处调用这个方法
     
     - parameter keys: key的数组
     */
    func addKeysTask(keys:NSArray) {
        self.taskContent=keys as Array<AnyObject>
    }
    
    /**
     添加多个事件对象的监听,在事件的触发处调用这个方法
     
     - parameter items: 模型的数组
     */
    func addItemsTask(items:NSArray) {
        self.taskContent=items as [AnyObject]
    }
    
    var maxTasks=30
    var taskContent:Array=[AnyObject]()
    var delegate:RunloopEventCaptureDelegate?
    
    func resgistRunloopWork(obj:RunloopEventCapture) {
        
        self.registerObserver(activities: CFRunLoopActivity.beforeWaiting.rawValue, order: INTMAX_MAX-999, mode: CFRunLoopMode.defaultMode , info: unsafeBitCast(obj, to: UnsafeMutableRawPointer.self)) { (runLoopObsver, activity, info) in
            
            RunloopEventCapture.sharedInstace.runLoopWorkDistributionCallback(observer: runLoopObsver!, activity: activity, info: info!)
        }
        
    }
    
    func runLoopWorkDistributionCallback(observer:CFRunLoopObserver , activity:CFRunLoopActivity , info:UnsafeRawPointer) {
        
        //print(":::",printActivity(activity));
        
        if (self.taskContent.count == 0) {
            return;
        }
        
        var result:Bool? = false
        while (result == false && self.taskContent.count != 0) {
            //保证我们需要的截获事件之执行一次
            if self.captureObjct == nil {
                break
            }
            
            let isContained:Bool=self.taskContent.contains(where: { (Obj) -> Bool in
                
                if Obj.isEqual(self.captureObjct){
                    return true
                }
                
                return false
            })
        
            if (isContained) {
                
                weak var weakSelf:RunloopEventCapture!=self
                DispatchQueue.global().async {
                    result=weakSelf.delegate?.asyncUploadCaptureUserData!(data: weakSelf.captureObjct!)
                    DispatchQueue.main.async {
                        self.captureObjct=nil
                    }
                }
                
                
                break
            }
        }
    }
    
    func registerObserver(activities:CFOptionFlags, order:CFIndex, mode:CFRunLoopMode, info:UnsafeMutableRawPointer, callback:@escaping CFRunLoopObserverCallBack ) {
        let runLoop:CFRunLoop=CFRunLoopGetCurrent()
        var context:CFRunLoopObserverContext=CFRunLoopObserverContext.init(version: 0, info: unsafeBitCast(self, to: UnsafeMutableRawPointer.self), retain: nil, release: nil, copyDescription: nil)
        let observer:CFRunLoopObserver=CFRunLoopObserverCreate(nil, activities, true, order, callback, &context)
//        RunLoop
        CFRunLoopAddObserver(runLoop, observer, mode)
    }
    
    //runloop的状态变化
    func printActivity(activity:CFRunLoopActivity) ->String{
        var activityDescription:String=""
        switch (activity) {
        case CFRunLoopActivity.entry:
            activityDescription = "kCFRunLoopEntry"
            break;
        case CFRunLoopActivity.beforeTimers:
            activityDescription = "kCFRunLoopBeforeTimers"
            break;
        case CFRunLoopActivity.beforeSources:
            activityDescription = "kCFRunLoopBeforeSources"
            break;
        case CFRunLoopActivity.beforeWaiting:
            activityDescription = "kCFRunLoopBeforeWaiting"
            break;
        case CFRunLoopActivity.afterWaiting:
            activityDescription = "kCFRunLoopAfterWaiting"
            break;
        case CFRunLoopActivity.exit:
            activityDescription = "kCFRunLoopExit"
            break;
        default:
            break;
            
        }
        return activityDescription
    }
    
}
